Use OrderContents to update line items

Matt-Yorkley opened the pull request:
#### What? Why?

- Consolidates all operations that **change the line items in an order** into a single service (`Spree::OrderContents`). This abstraction was actually added in Spree 2.0, but we mostly ignored it instead of implementing it. :see_no_evil: 

- Removes the call to `order.update!` from the LineItem after_save callback. This means if we ever update 20 line items at once (which is not uncommon), we will call `order.update!` only once instead of calling it 20 times. Even calling it once is expensive!

- Refactors places touched by `OrderContents` changes, including `CartController#populate`, which currently is one of our worst offenders for performance. There are some changes here that make the code clearer and fix some nasty issues.

- Fixes a bug that meant editing orders via BOM could shipping fees and taxes in an incorrect state... :open_mouth: 

- After this, if we need to make changes to the logic around what things are updated when an order's line items change (updating adjustments, taxes, order totals, etc) we can make those changes in one place instead of a dozen places.

**TLDR version:** 

From now on, any time we want to add/edit/remove some line items on an order, we _tell_ `OrderContents` what we want to do, and `OrderContents` will do it and make sure _all the things that need to be done when the order's line items change_ get done, instead of that logic being spread all over the app (in each individual place where we change line items).

:+1:

#### What should we test?
<!-- List which features should be tested and how. -->

Mostly we should test adding and removing items from the cart (in the shopfront page) and try really hard to break it. Also check performance on the basket page when updating orders with lots of line items. It should be improved.

Other places that have been touched include BOM and editing line items in and order in admin. They should work as before. We recently added additional test coverage in both of those places, so I think we can be fairly confident.

#### Release notes
<!-- Write a one liner description of the change to be included in the release notes.
Every PR is worth mentioning, because you did it for a reason. -->

Consolidated line-item-changing logic into OrderContents service.

<!-- Please select one for your PR and delete the other. -->
Changelog Category: Technical changes

luisramos0 reviewed (commented):
Very good one Matt!
This is the natural continuation of a refactor I did after moving the code from Spree. The populator was a mess. I remember I tried and failed to use OrderContents to add line items to the order in the frontoffice. Basically, I would have to make a Pr like this one. Quite a lot of changes :clap: :clap: :clap: 
This is really good progress in the right direction.

My comments:
- is there anything we need to verify in terms of variant overrides in these changes?
- in the tests I'd mix adding line items in the frontoffice with changes to the order made in the backoffice to verify that they both work well together

Matt-Yorkley commented:
> is there anything we need to verify in terms of variant overrides in these changes?

I don't think there are any changes here in relation to how we're already doing (or not doing) variant scoping. It's obscure though! In terms of line items, the main attributes from a variant override are `price` and `on_hand`. In the cases where it's relevant, the values being passed from the frontend is being overridden and the stock check is handled. The main two are `CartService` and `Api::ShipmentController`. In some cases where we're incrementing or decrementing an existing line item's quantity, the `price` attribute has already previously been set (for example) based on an override. There were places where I was shocked that we were not doing variant scoping, but they had specific variant override tests, and they were passing! :man_shrugging: 

I definitely want to add some more tests in a couple of key places :100:

Matt-Yorkley commented:
So when we save a line item we validate it with our modified version of `Spree::Stock::AvailabilityValidator` here:
https://github.com/openfoodfoundation/openfoodnetwork/blob/cea3ee4ef9a852936c61585ecd0875ab70f52340/app/models/spree/line_item.rb#L32
Which uses scoped variants here:
https://github.com/openfoodfoundation/openfoodnetwork/blob/cea3ee4ef9a852936c61585ecd0875ab70f52340/app/models/spree/stock/availability_validator.rb#L40

So I think generally we're pretty safe when it comes to updating existing line items, for example in places like the Bulk Order Management page. If the stock is not available (including overrides) it won't save the changes.

The main places we create line items are `CartService` and `Api::ShipmentController` and they are both covered with variant override specs (and both use variant overrides very explicitly), eg:
https://github.com/openfoodfoundation/openfoodnetwork/blob/cea3ee4ef9a852936c61585ecd0875ab70f52340/spec/controllers/api/v0/shipments_controller_spec.rb#L74
https://github.com/openfoodfoundation/openfoodnetwork/blob/cea3ee4ef9a852936c61585ecd0875ab70f52340/spec/features/consumer/shopping/variant_overrides_spec.rb#L77

I can't see any specific areas that need more coverage..?

Matt-Yorkley commented:
Adding more specs uncovered a pretty nasty issue which is fixed by this PR: https://github.com/openfoodfoundation/openfoodnetwork/pull/7623/commits/ee25f6b7a43cb2199699a0239e72d28d0b5dfe97

luisramos0 commented:
Xit :-)

On Mon, 17 May 2021, 14:04 Matt-Yorkley, ***@***.***> wrote:

> Adding more specs uncovered a pretty nasty issue which is fixed by this
> PR: ee25f6b
> <https://github.com/openfoodfoundation/openfoodnetwork/commit/ee25f6b7a43cb2199699a0239e72d28d0b5dfe97>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openfoodfoundation/openfoodnetwork/pull/7623#issuecomment-842306970>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMQPOVHLYCXNM3TMROU6PDTOEH6BANCNFSM44ZAJ5AA>
> .
>

Matt-Yorkley commented:
Alright, I realise this PR is a bit of a monster but... I think we probably need to do this big `OrderContents` refactor before we can finish adjustments, and we'll also need it if we ever want to escape **callbacks hell**.

luisramos0 commented on app/models/spree/order.rb:
```diff
@@ -517,6 +465,21 @@ def create_proposed_shipments
       shipments
     end
 
+    # Clear shipments and move order back to address state
```
there's an if missing in this comment:
"Clear shipments and move order back to address state if order is not complete but contains shipments" ?

luisramos0 commented on app/models/spree/order.rb:
```diff
@@ -517,6 +465,21 @@ def create_proposed_shipments
+        state: checkout_steps.first,
+        updated_at: Time.zone.now,
+      )
+    end
```
hmm, I think this is https://refactoring.guru/smells/feature-envy this restart_checkout_flow should go to service OrderCheckoutRestart or the module Order::Checkout itself. order.rb shouldnt have order state logic I think.

Matt-Yorkley replied:
This bit was pulled straight from Spree 2.2, but yes it probably wants to be somewhere else :ok_hand:

Matt-Yorkley replied:
Moved :+1:

luisramos0 commented on app/models/spree/order_contents.rb:
```diff
@@ -11,25 +11,79 @@ def initialize(order)
+        true
+      else
+        false
+      end
```
I think these methods taking params are clearly breaking the single responsibility principle, a class should do one thing. What if you create a separate class for this? Or even two classes: one to update order based on params, another to update a line item based on params... you will know better.

Matt-Yorkley replied:
So this is another bit pulled straight from Spree 2.2. It's moved here to keep this logic out of an admin and an API controller, but it's used where the params are a hash of line items data (with `accepts_nested_attributes`).

Matt-Yorkley replied:
So the params are basically:
```
{order:
  line_items_attributes: [
    id: 1, quantity: 5,
    id: 4, quantity: 8,
    id: 6, quantity: 9
  ]
}
```
and it's more of a bulk-update of the order's line items than an update of the order...

Matt-Yorkley replied:
So the responsibility here is still roughly "performing operations that change line items, including any updates required as part of those changes".

luisramos0 commented on app/models/spree/order_contents.rb:
```diff
@@ -11,25 +11,79 @@ def initialize(order)
 
-      unless line_item
-        raise ActiveRecord::RecordNotFound, "Line item not found for variant #{variant.sku}"
+    def update_or_create(variant, attributes)
```
I wonder about the name of the method and the interface of this class. could this be update_or_add or is creating and adding different?
could we rename add to add_variant and remove to remove_variant and here we could rename to update_or_create_line ?
together with the methods taking params below this interface is quite complicated right now.

I understand the PR is basically moving it all to this one class, which is great! I am now suggesting we organize the code that has landed here :-)

Matt-Yorkley replied:
This bit wasn't from Spree. It's where I had to do some adapting to make `OrderContents` work with OFN's `CartService`: https://github.com/openfoodfoundation/openfoodnetwork/pull/7623/commits/dd34c0708f7e1fc5eeb804b81719acaec10ae4d0

it's either creating or updating a line item from a given variant, but it's a bit messy due to that extra complexity in `CartService`.

Matt-Yorkley replied:
Yeah there's a lot more that can be done here! I struggled to stop at under 30 commits, I could have carried on and done 40+.

luisramos0 reviewed (approved):
I think this PR is pretty nice Matt, I personally dont mind the size in this case.
I just left some comments about code structure. Please ignore them if you will, I know it's complicated!

andrewpbrett commented on spec/controllers/admin/bulk_line_items_controller_spec.rb:
```diff
@@ -209,10 +209,10 @@
             expect(line_item1.order).to receive(:update_line_item_fees!)
             expect(line_item1.order).to receive(:update_order_fees!)
-            expect(line_item1.order).to receive(:update_order!).twice
+            expect(line_item1.order).to receive(:update_order!).once
```
💪

andrewpbrett commented on spec/controllers/admin/bulk_line_items_controller_spec.rb:
```diff
@@ -373,21 +376,22 @@
-        expect(order.total).to eq 22.0
+        expect(order.total).to eq 19.0
+        expect(order.shipment_adjustments.shipping.sum(:amount)).to eq 3.0
+        expect(order.shipment_adjustments.tax.sum(:amount)).to eq 0.14
         expect(order.item_total).to eq 10.0
-        expect(order.adjustment_total).to eq 12.0
-        expect(order.included_tax_total).to eq 0.99
+        expect(order.adjustment_total).to eq 9.0
+        expect(order.included_tax_total).to eq 0.84
```
Are these the changes to the totals that correct the bug that's currently in production?

Matt-Yorkley replied:
Shockingly, it shows an entirely separate bug which wasn't even documented (which is fixed here). The shipping fee was not being updated correctly, but it is now.

andrewpbrett reviewed (approved):
This looks like a really nice improvement, it is pretty huge though! I focused on whether the changes to the specs make sense and I think they do. So combining the green build with some thorough testing should give us good confidence to merge.

filipefurtad0 commented:
Hey @Matt-Yorkley,

I've mostly ran three test to check this huge work (below). Also smoke tested the editing of the editing of orders. Awesome that you've spotted the issue with order editing and shipping fees! :muscle: 


**Editing the Cart**

The two videos below show the editing of the quantity of the each of the 14 line items in the `/cart` page. It seems there are more ActiveRecord access calls, before staging :+1: 

Before staging:

![Peek 2021-06-04 13-41](https://user-images.githubusercontent.com/49817236/120802805-c2afac80-c53a-11eb-854a-e56cc421ff13.gif)


After staging:
![Peek 2021-06-04 13-07](https://user-images.githubusercontent.com/49817236/120801996-d1e22a80-c539-11eb-875c-237f3848c3d9.gif)


**Editing the Order (Backoffice)**

Before staging - when editing an order, shipping fees are not updated automatically. The outcome is the same when this is performed through BOM.

![Peek 2021-06-04 12-53](https://user-images.githubusercontent.com/49817236/120802087-ed4d3580-c539-11eb-9f8f-cdd06262689e.gif)


After staging - shipping fees are updated when adding a new line item :+1:  but if the item is removed the shipping fee should be removed as well :-1: 

![Peek 2021-06-04 13-00](https://user-images.githubusercontent.com/49817236/120804703-e07e1100-c53c-11eb-9c07-14abf4fda8af.gif)

In both cases, the shipping fee can be updated manually, but I guess it should work automatically, without the need to click Update and Recalculate Fees, right?


**Adding Items to the Cart**
Hum, something strange is happening here. The number of items added to the cart differs, between the /shopfront and /cart page. This is reproducible only after staging this PR.

Before staging:

![Peek 2021-06-04 13-27](https://user-images.githubusercontent.com/49817236/120801232-d1955f80-c538-11eb-8205-01c63444c719.gif)

After staging:

![Peek 2021-06-04 13-17](https://user-images.githubusercontent.com/49817236/120801219-ce01d880-c538-11eb-83df-12cbe28a35f2.gif)

I think this last issue is a bit critical. It's reproducible after staging this PR. Basically, some items seem to get removed, when proceeding to the cart page. This looks like a blocker for merging.

filipefurtad0 commented:
Perhaps worth adding one extra thought, to stress out the performance improvement.

The tests above were comparing this PR with the current master; the current master has changed quite a bit, with the merging from Rails 6.

This is how the same test looks like in the french server, previous to the merging of Rails 6.0:

![Peek 2021-06-04 15-40](https://user-images.githubusercontent.com/49817236/120821247-33f85b00-c54d-11eb-84d3-68692251a81d.gif)

Matt-Yorkley commented:
> some items seem to get removed, when proceeding to the cart page

Thanks! I added an extra change to try to improve the efficiency of the cart processing but I wasn't 100% sure about it, which is what the "try really hard to break it" testing note was about :sweat_smile: I've removed that specific bit, it should be back to normal.

filipefurtad0 commented:
This is looking great now @Matt-Yorkley,

I've tried the same scenario: adding plenty of items and proceeding to the cart page. After your latest commits this seems to be fixed, the item number is consistent between /cart and /shop page.

For documentations sake:

![Peek 2021-06-05 10-38](https://user-images.githubusercontent.com/49817236/120887545-6e183a00-c5eb-11eb-9ddc-7a352bde8ded.gif)

The bug fix is also great. There is just the case in which the line item is removed and a shipping fee remains (on the previous notes). But I think this can be dealt with in a separate issue, as it is not introduced here.

Moving to ready to go :tada:
